l1-contracts/contracts/bridge/ntv/L2NativeTokenVault.sol

```diff
@@ -91,32 +91,32 @@ contract L2NativeTokenVault is IL2NativeTokenVault, NativeTokenVault {
     }
 
     /// @notice Ensures that the token is deployed.
-    /// @param _originChainId The chain ID of the origin chain.
     /// @param _assetId The asset ID.
     /// @param _originToken The origin token address.
     /// @param _erc20Data The ERC20 data.
     /// @return expectedToken The token address.
     function _ensureTokenDeployed(
-        uint256 _originChainId,
         bytes32 _assetId,
         address _originToken,
         bytes memory _erc20Data
     ) internal override returns (address expectedToken) {
-        expectedToken = _assetIdCheck(_originChainId, _assetId, _originToken);
+        uint256 tokenOriginChainId;
+        (expectedToken, tokenOriginChainId) = _calculateExpectedTokenAddress(_originToken, _erc20Data);
         address l1LegacyToken;
         if (address(L2_LEGACY_SHARED_BRIDGE) != address(0)) {
             l1LegacyToken = L2_LEGACY_SHARED_BRIDGE.l1TokenAddress(expectedToken);
         }
 
         if (l1LegacyToken != address(0)) {
-            /// token is a legacy token, no need to deploy
-            if (l1LegacyToken != _originToken) {
-                revert AddressMismatch(_originToken, l1LegacyToken);
-            }
-            tokenAddress[_assetId] = expectedToken;
+            _ensureTokenDeployedInnerLegacyToken({
+                _assetId: _assetId,
+                _originToken: _originToken,
+                _expectedToken: expectedToken,
+                _l1LegacyToken: l1LegacyToken
+            });
         } else {
             super._ensureTokenDeployedInner({
```

Comment:
- `expectedToken` in this function will be an address get from `_calculateExpectedTokenAddress`.
- Assume 
        - `tokenOriginChainId` is L1 ChainId
        -  the `L2NativeTokenVault` is deployed on Era
        - It is a first deposit for a token 
- then `_calculateExpectedTokenAddress` call shared bridge `l2TokenAddress` to get expected token address.
- `L2SharedBridge` callback `L2NativeTokenVault:l2TokenAddress` get zero address and so the `expectedToken` will be calculated as `L2SharedBridge:_calculateCreate2TokenAddress`
- `_ensureTokenDeployedInner` will be called in `_ensureTokenDeployed` because `l1LegacyToken == address(0)`
- Then when `_deployBridgedToken` will be called it will generate salt as `_getCreate2Salt` which has formula of 

```
    function _getCreate2Salt(uint256 _originChainId, address _l1Token) internal view virtual returns (bytes32 salt) {
        salt = keccak256(abi.encode(_originChainId, _l1Token));
    }
```

While `expectedToken` was generated with 

```
    function _getCreate2Salt(address _l1Token) internal pure returns (bytes32 salt) {
        salt = bytes32(uint256(uint160(_l1Token)));
    }
```

salt